### ISSUE TITLE
Python installation through CMake improvements

### ISSUE THREAD

**Issue by brenthuisman** [arbor-sim/arbor#1403]
Three ways to improve our Python installation when the user uses CMake:

* [ ] `python setup.py egg_info` generates an `arbor.egg-info` dir, which should be copied along with `arbor` when `make install` copies to `site-packages`. This will register the package to pip, and allows pip to manage it.
* [x] We should NOT be overriding the install destination (obtained with `sysconfig.get_path('platlib')`) when a CMAKE_INSTALL_PREFIX is provided. This ignores pip/conda environments, and is thus unexpected behavior for Python users. Destination for C++ artifacts is entirely separate for Python folks, but still may be needed (modcc).
* [ ] ARB_USE_BUNDLED_LIBS should default to ON. The default should be to go with our vetted and provided versions of dependencies.
  * Related: we don't check nonbundled deps such as `pybind` for versions. Should we? I'd like to avoid users [successfully compiling broken packages](https://github.com/arbor-sim/arbor/issues/1398).

**Comment by bcumming** [comment]
This might want to be three different questions!

My overall feedback here is that the `pip install arbor` process follows your suggestions for `INSTALL_PREFIX` and `BUNDLED_LIBS`, to give the standard Python experience. On the other hand, when using CMake, the installation should follow standard packaging rules and expectations, not what our hypothetical Python user expects.

* Regarding eggs, I am showing the limits of my knowledge here: but I thought that eggs were the "old way", and we should use wheels instead?

* Disagree there: if an explicit CMAKE_INSTALL_PREFIX is set, the standard behavior is to install everything in that path. If a user, be they Python or not, is surprised, it doesn't change the fact that this is where they asked to install. For example, if we want to make our installation work in automated environments, it has to use `--prefix` paths like inteneded. e.g. I haven't checked what happens with spack, but I expect that it expects the package to be installed following those rules.
One option for dealing with this is to print a message during CMake configuration that indicates when the requested installation path does not match `sysconfig.get_path('platlib')`.

* Let's start checking the versions of the dependencies (specifically pybind11 and json). There are safety reasons for using the system version, that will lead to even nastier bugs when they pop up.

**Comment by bcumming** [comment]
Is the `egg-info` issue related to the problems we have with pip not being able to unistall the arbor package?

**Comment by brenthuisman** [comment]
Python users (which includes people who `ARB_WITH_PYTHON=ON`) expect certain standard packaging rules and expectations, which is to obey `sysconfig.get_path('platlib')`. This is not hypothetical.

1. Apparently this is still how `pip` picks up packages in its registry. Packges without a companion 'egg' directory cannot be managed by `pip`, and do not show up.
2. Setting `CMAKE_INSTALL_PREFIX` need not imply more than that the user is (rightly so) trained to not use `sudo` when not managing system functionality. There is simply no way to `make install` the Python package in the current `env` _and_ install the C++ artifacts somewhere in `~` without using sudo or manual config. We should either decouple (have a separate configurable prefix for Python) or discouraged using `CMAKE_INSTALL_PREFIX` to avoid using `sudo` by defaulting the `CMAKE_INSTALL_PREFIX` to `~/.local`.
3. Should we test these? As seen in the linked bug report, even successful compilation is no guarantee. In any case, the default should be (perhaps contingent on setting `ARB_WITH_PYTHON=ON`) to use bundled libs.

**Comment by Helveg** [comment]
@bcumming Yes: eggs are old, wheels are good, so mea culpa, the command should be `python setup.py dist_info` to generate the wheel's `.dist-info` folder instead of the `.egg-info`. Also yes: The lack of metadata makes the `arbor` folder entirely invisible to all of the python package tools like `pip`, `setuptools`, `pkg_resources` and distutils` and they are the sources of truth on all that is packages in Python. The `dist-info` folder contains some simple `.txt` files locating all of the files of the package, specifying metadata like authors and version blablabla.

I generally don't think it makes sense to create Python bindings, force Python to **discover them as importable** (be it through `PYTHONPATH` or by sneaking them in the packages folder) but not inform the Python packaging system.

Your second point about the `DCMAKE_INSTALL_PREFIX` is a good point, if it's conventional to containerize under the install prefix, do so, but Pythonistas _need_ a way to leverage the full CMake capabilities while very explicitly asking for a full fledged Python package as end result. I'd be fine and settle for a `-DWITH_PYTHON_PACKAGE` flag, or a `-DWITH_ADMIT_YOURE_A_DIRTY_PYTHONISTA` flag :)

**Comment by Helveg** [comment]
@bcumming After some digging it turns out that the `dist-info` folder is a part of the `wheel` spec and wheels don't provide a stable public API. The relevant code is

```python
        distinfo_dirname = '{}-{}.dist-info'.format(
            safer_name(self.distribution.get_name()),
            safer_version(self.distribution.get_version()))
        distinfo_dir = os.path.join(self.bdist_dir, distinfo_dirname)
        self.egg2dist(self.egginfo_dir, distinfo_dir)
```

This located at the tail end of the `bdist_wheel` command to use the build information to transform the `egg-info` folder into a `dist-info` folder.

Now, using `python setup.py dist_info` only generates part of the metadata required and doesn't result in a succesful uninstall. Using `python setup.py egg_info` yields better (actually flawless) results:

```
(arb-cat) robin@TNG2019:~/git/arbor$ pip show arbor
WARNING: Package(s) not found: arbor
(arb-cat) robin@TNG2019:~/git/arbor$ python setup.py egg_info
/home/robin/.pyenv/versions/arb-cat/lib/python3.9/site-packages/setuptools/dist.py:458: UserWarning: Normalizing '0.5.1-dev' to '0.5.1.dev0'
  warnings.warn(tmpl.format(**locals()))
running egg_info
writing arbor.egg-info/PKG-INFO
writing dependency_links to arbor.egg-info/dependency_links.txt
writing requirements to arbor.egg-info/requires.txt
writing top-level names to arbor.egg-info/top_level.txt
reading manifest file 'arbor.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'arbor.egg-info/SOURCES.txt'
(arb-cat) robin@TNG2019:~/git/arbor$ mv arbor.egg-info/ /home/robin/.pyenv/versions/arb-cat/lib/python3.9/site-packages/
(arb-cat) robin@TNG2019:~/git/arbor$ pip show arbor
Name: arbor
Version: 0.5.1.dev0
Summary: High performance simulation of networks of multicompartment neurons.
Home-page: https://github.com/arbor-sim/arbor
Author: The Arbor dev team.
Author-email: None
License: UNKNOWN
Location: /home/robin/.pyenv/versions/3.9.1/envs/arb-cat/lib/python3.9/site-packages
Requires: numpy
Required-by:
(arb-cat) robin@TNG2019:~/git/arbor$ pip uninstall arbor
Found existing installation: arbor 0.5.1.dev0
Uninstalling arbor-0.5.1.dev0:
  Would remove:
    /home/robin/.pyenv/versions/3.9.1/envs/arb-cat/lib/python3.9/site-packages/arbor
    /home/robin/.pyenv/versions/3.9.1/envs/arb-cat/lib/python3.9/site-packages/arbor-0.5.1.dev0-py3.9.egg-info
Proceed (y/n)? y
  Successfully uninstalled arbor-0.5.1.dev0
```

It seems like some metadata is missing from `setup.py` like "Author-email" and "License" but other than that it seems to be treated as a Python package just by running `python setup.py egg_info` and `mv arbor.egg-info /to-sitepackages`

**Comment by bcumming** [comment]
I think that there is one remaining contentious point, which is the default installation directory.

The rule when installing is to install the package where in the `--prefix` / `CMAKE_INSTALL_PREFIX`. You _do not_ override that path with arbitrary state from whatever version of Python/pip/conda the user has configured.

At the moment a user of CMake can install to their virtual env without sudo, by providing the path. E.g. 

`cmake -DCMAKE_INSTALL_PREFIX=/path/to/virtual-env`. That will:
- create the appropriate sub-directory (`lib/python3.9/site-packages/` on my system) and copy the python module there.
- install all the docs, arbor lib, headers, etc in the relevant `include`, `lib`, `doc` etc paths, for a nicely packaged installation in your venv.

We can document this better in the docs. Put a big warning box around it, add flashing lights, whatever it takes to catch people's attention.

We could also add an extra flag, e.g. `-DARB_PYTHON_INSTALL_ENV_PREFIX=on` (I know, we could find something less verbose).

**Comment by Helveg** [comment]
`cmake -DCMAKE_INSTALL_PREFIX=/path/to/virtual-env` loses the ease of use that `/path/to/virtual-env` is derived from the active python interpreter. Does it not? But the flag would fix that?

And uhm, are you sure that setting the `CMAKE_INSTALL_PREFIX` to `site-packages` is equivalent to placing the Python bindings and package metadata there?

That would imply 2 things that I doubt but will test:

* Python automatically discovers `/path/to/site-packages/arbor/lib/python3.9/site-packages/arbor` in exactly the same way as it would discover `/path/to/site-packages/arbor`. I do not know of any such deep digging of Python into folders. My knowledge is that the import machinery of Python treats everythy Python file on its `PATH` as a module and every folder with an `__init__.py` similarly as a module. I don't know of any default `lib/python<ver>/site-packages` explorations of things already inside of `site-packages`
* The package metadata is somewhere in the CMake generated install code?

I'm prepared to be happily surprised

**Comment by brenthuisman** [comment]
To be clear: what @bcumming propose is not how it works currently. CMAKE_INSTALL_PREFIX defaults to `/usr/local`, and the Python install path to the output of `sysconfig.get_path('platlib')`.

Once the user sets CMAKE_INSTALL_PREFIX, those two are coupled, unless the user _also_ sets ARB_PYTHON_LIB_PATH. There is no flag to leave ARB_PYTHON_LIB_PATH at the `sysconfig.get_path('platlib')` and specify a custom CMAKE_INSTALL_PREFIX, so a user has to query Python themselves to obtain the right `site-packages` for their env again. Not good UX.

CMAKE_INSTALL_PREFIX defaulting to `sysconfig.get_path('platlib') ../../` could be a nice alternative to `ARB_PYTHON_RESPECT_CMAKE_INSTALL_PREFIX=on`.

**Comment by Helveg** [comment]
Setting `-DCMAKE_INSTALL_PREFIX` to the root of my venv `/home/robin/.pyenv/versions/3.9.1/envs/arb-test` with:

```
cmake -DCMAKE_INSTALL_PREFIX=/home/robin/.pyenv/versions/3.9.1/envs/arb-cat -DARB_WITH_PYTHON=ON -DARB_USE_BUNDLED_LIBS=ON ..
```

doesn't really do what could be improved: 

* [x] The bindings end up in the site-packages
* [ ] The install prefix is derived from my Python interpreter
* [ ] The package metadata is generated and `pip show arbor` and `pip uninstall arbor` work

Ideally something like: `cmake -DARB_PYTHON_INSTALL=ON -DARB_USE_BUNDLED_LIBS=ON ..` derives the `CMAKE_INSTALL_PREFIX` from the Python interpreter, does what `DARB_WITH_PYTHON=ON` does and checks the 2 unchecked marks above

**Comment by bcumming** [comment]
@brenthuisman: `CMAKE_INSTALL_PREFIX` defaults to whatever the system default is (99% likely `/usr/local`). What I was saying is that  the user can manually set it to their target virtualenv path (or whatever Python specific target they have in mind), which @Helveg demonstrated.

By default, CMAKE_INSTALL_PREFIX has to stay at its default value, and if it is set by the user, everything has to get installed there, so that Arbor plays by the rules expected by every OS package manager, spack, easybuild, modules on your HPC system, and so on.

You are right that it is a bit fiddly if the user wants to set different CMAKE_INSTALL_PREFIX and ARB_PYTHON_PREFIX paths, but some level of complexity will arise when a user wants to customise "all the paths".

W can improve the UX for users wanting to install Arbor, and I think that there are two constraints:
- It must be opt in
- We don't want to end up with a lot of `if-else` or distributed logic in CMake: long term maintenance issues ahoy!

How do we make the opting in easier? some ideas:
- provide a flag like `ARB_PYENV_INSTALL_PREFIX=on` which would force `CMAKE_INSTALL_PREFIX` to be python specific
  - this will require additional logic to error out if the user has also set `CMAKE_INSTALL_PREFIX`.
  - this has the potential to create some other logic bombs in the CMake
 - we could try to make a special target like `make python-install` that installs at whatever we get from the Python env.
   - `make install` installs everything in `CMAKE_INSTALL_PREFIX` like "normal"

**Comment by bcumming** [comment]
@Helveg, happy to hear that setting the path worked for you. Ir response to your unfilled check boxes:

* **The install prefix is derived from my Python interpreter** I suppose that isn't surprising, given that you manually provided the installation path. 
* **The package metadata is generated and pip show arbor and pip uninstall arbor work.** and **The arbor module is respected and not overwritten by pip when installing packages with arbor as dependency.**: how much do we want to pursue these aspirations, given that we didn't install Arbor using pip? If we make this work out of the box for pip, will it also work for Conda, or your fave package manager that isn't pip?

**Comment by Helveg** [comment]
1) Ofcourse, but I'd like a workflow where I don't have to. I want to indicate that I want to use CMake as a build system to build Arbor as a Python package for my active Python interpreter; a valid and probably common use case for Python devs that want to dig a little deeper into arbor, but still want it treated as a pip installed package.
2) Yes, as these metadata specifications are outlined in [PEP314](https://www.python.org/dev/peps/pep-0314/) and are the law for all package managers. The issues between `pip` and `conda` lie not in the metadata specification but in the **build systems** they provide for building C/C++ extensions, and again that this is completely covered by CMake in this workflow.

**Comment by halfflat** [comment]
I have to assert that safety and predictability are paramount when it comes to installing under a prefix. CMake already plays by its own rules when it comes to installation options — there's no way, really, to have it be configured for one install path, but copied to another, as one would want for staging or use with `stow` — but writing outside the install prefix is just asking for trouble. It makes it hard to isolate, hard to rollback, surprising to systems administrators, and generally speaking, means the package is a bad Unix citizen.

We should definitely make pip installs as straightforward as possible, and if there is a way to help those building and installing from source get the python module where they would most likely want it, that's also great. But not at the expense of violating the expectation that everything is installed under the given install prefix, if provided.

**Comment by Helveg** [comment]
@halfflat I don't think any of my proposals require us to misbehave and can all function perfectly under the `CMAKE_INSTALL_PREFIX`, the only reason I initially thought I had to is because I had no idea I could just target my `venv` as `CMAKE_INSTALL_PREFIX`!

**Comment by brenthuisman** [comment]
@Helveg do you know what the difference is between egg_info and dist_info? [This page](https://packaging.python.org/discussions/wheel-vs-egg/) very briefly discusses egg_info in the past tense; [here](https://setuptools.readthedocs.io/en/latest/userguide/commands.html) eggs are warned against for being deprecated. I can't find a clear source stating that dist_info is the new way.

**Comment by Helveg** [comment]
See https://www.python.org/dev/peps/pep-0427/#the-dist-info-directory.

`dist-info` is part of the wheel format and is an extension on the `egg-info`, which iirc contains the egg info completely. While eggs might be deprecated, the "Metadata for Python Software Packages" spec continues on (into [v2.2](https://www.python.org/dev/peps/pep-0639/) atm I believe) and setuptools produces compliant `.egg-info` folders.

If you take a look at the source code of the `wheel` package you'll see that they also rely on `egg_info` to create their `dist-info` as `distutils` and `setuptools` are still the stdlib sources of truth for packaging in Python (and let's hope they follow their own spec)

**Comment by brenthuisman** [comment]
@Helveg Thanks for the info! On @bcumming's machine the distro (Arch) seemed to accompany `dist_info` dirs for system installed packages (to `/usr`), while on my distro (Fedora) I could see `egg_info`s. That made us unsure what, if any, is correct at this point, or that apparently it's up to distros to choose. Dealing with this correctly requires more logic/magic in the CMake install, while `python3 setup.py install` will use setuptools and therefore always do what's the systems norm. The linked PR therefore simply stresses that if you care about proper Python/pip integration (respecting envs etc), `setup.py` is what you're looking for.

tldr: It makes more sense to solve any problems with `setup.py` (installing modcc etc) than trying to mimic setuptools in our CMake and inevitably run into subtle or not so subtle differences between distros.

**Comment by bcumming** [comment]
We chose the default of `ARB_USE_BUNDLED_LIBS=off` according to the principle of least surprise.
If a user has configured their environment to make a version of json or pybind11 available, they will be surprised if it isn't detected by default.

**Comment by brenthuisman** [comment]
I do not recall the discussion, but I'm willing to blame my memory.

First point (generating the `egg_info`) is still open in any case.

**Comment by bcumming** [comment]
You weren't involved in the conversation (It was @halfflat, @schmitts and I).
But I think I made that point higher up on this ticket.
If you really want to keep pushing for it, we should separate it from the discussion about `egg-info`, which I am in favor of.

**Comment by brenthuisman** [comment]
The concern is that users will be compiling combinations that we have never tested. Now, these libraries are quite stable, so maybe someone can reassure me by saying problems due to this are unlikely. If not, is it wise to leave all Spack users with untested versions of Arbor? Is there any testing infrastructure for updates to Spack packages?

**Comment by brenthuisman** [comment]
Stale and solved by advertising pip as preferred installation method for users in Python ecosystems.
